go-controller/pkg/util/pod_annotation.go

```diff
@@ -0,0 +1,221 @@
+//   annotations:
+//     k8s.ovn.org/pod-networks: |
+//       {
+//         "default": {
+//           "ip_address": "192.168.0.5/24",
+//           "mac_address": "0a:58:fd:98:00:01",
+//           "gateway_ip": "192.168.0.1"
+//         }
+//       }
+//
+// (With optional additional "routes" also indicated; in particular, if a pod has an
+// additional network attachment that claims the default route, then the "default" network
+// will have explicit routes to the cluster and service subnets.)
+//
+// In a dual-stack cluster, the annotation looks like:
+//
+//   annotations:
+//     k8s.ovn.org/pod-networks: |
+//       {
+//         "default": {
+//           "ip_addresses": ["192.168.0.5/24", "fd01::1234/64"],
+//           "mac_address": "0a:58:fd:98:00:01",
+//           "gateway_ips": ["192.168.0.1", "fd01::1"]
+//         }
+//       }
+//
+// Note that the public PodAnnotation type always has an array of IPs and Gateways,
+// regardless of whether the cluster is single or dual-stack.
+
+const (
+	// OvnPodAnnotationName is the constant string representing the POD annotation key
+	OvnPodAnnotationName = "k8s.ovn.org/pod-networks"
+	// OvnPodDefaultNetwork is the constant string representing the first OVN interface to the Pod
+	OvnPodDefaultNetwork = "default"
+)
+
+// PodAnnotation describes the assigned network details for a single pod network. (The
+// actual annotation may include the equivalent of multiple PodAnnotations.)
+type PodAnnotation struct {
+	// IPs are the pod's assigned IP addresses/prefixes
+	IPs []*net.IPNet
+	// MAC is the pod's assigned MAC address
+	MAC net.HardwareAddr
+	// Gateways are the pod's gateway IP addresses; note that there may be
+	// fewer Gateways than IPs.
+	Gateways []net.IP
+	// Routes are additional routes to add to the pod's network namespace
+	Routes []PodRoute
+}
+
+// PodRoute describes any routes to be added to the pod's network namespace
+type PodRoute struct {
+	// Dest is the route destination
+	Dest *net.IPNet
+	// NextHop is the IP address of the next hop for traffic destined for Dest
+	NextHop net.IP
+}
+
+// Internal struct used to marshal PodAnnotation to the pod annotation
+type podAnnotation struct {
+	// Single/dual-stack version of IPs
+	IP  string   `json:"ip_address,omitempty"`
+	IPs []string `json:"ip_addresses,omitempty"`
+
+	MAC string `json:"mac_address"`
+
+	// Single/dual-stack version of Gateways
+	Gateway  string   `json:"gateway_ip,omitempty"`
+	Gateways []string `json:"gateway_ips,omitempty"`
+
+	Routes []podRoute `json:"routes,omitempty"`
+}
+
+// Internal struct used to marshal PodRoute to the pod annotation
+type podRoute struct {
+	Dest    string `json:"dest"`
+	NextHop string `json:"nextHop"`
+}
+
+// MarshalPodAnnotation returns a JSON-formatted annotation describing the pod's
+// network details
+func MarshalPodAnnotation(podInfo *PodAnnotation) (map[string]string, error) {
+	pa := podAnnotation{
+		MAC: podInfo.MAC.String(),
+	}
+
+	if len(podInfo.IPs) == 1 {
```

Comment:
@danwinship instead of this kind of if/else, should we have IP always equal IPs[0] if they are both set? Then here we just need to set IP = IPs[0] (same for gateway). Then when unmarshalling we could always look at IPs and fall back to IP?

Reply:
That's how I'd done it in an earlier version, but it felt ugly to be writing out an annotation with redundant `ip_address` and `ip_addresses` fields.

> eg at some point in the future I'd expect us to get rid of the singular fields entirely.

I wasn't necessarily assuming we'd do that... though I guess if we do want to do that then we should definitely write out both fields now.